### DISCUSSION TITLE
Ingress vs port forwarding

### DISCUSSION THREAD

**Discussion by josibake** [bitcoin-dev-project/warnet#529] [Ideas]
Starting a thread from https://github.com/bitcoin-dev-project/warnet/issues/526#issuecomment-2329439715:

> Might be worth a discussion on pros / cons on ingress.
>
> Downsides that come to mind is it's another small step to enable on minikube (just works on docker desktop) and that on cloud you get billed for public IP addresses and load balancers.
>
> Benefit of ingress is it allows anyone to connect to the same service which is nice and might be a bit more reliable than port-forwarding.

**Comment by josibake** [comment]
@m3dwards I had forgotten about needing load balancers, that's a good point w.r.t cost. Regarding Minikube, I don't _really_ care? if I'm dev'ing locally, port forward seems fine to me and running a load balancer seems like a PITA. But for a managed cluster, potentially with many users connecting, port forwarding seems less than ideal.

My main issues with port-forwarding:

- ive found it to be incredibly flakey , often failing and needing to be restarted
- its quite annoying when running warnet on a cloud provider from a headless environment. sure, I can do all this port forwarding and open ports on my devbox IP, but people able to deploy to GKE / DO and then just hit an endpoint is much nicer
- it forces us to push stuff through `warnet` commands so we can chain commands and do the port-forwarding for the user

Regarding cost, this would be good to quantify. If it's a fraction on top of what we are already paying for to run these clusters, seems fine? But it if ends up increasing total cost by like, 10%.. less ideal

**Reply by m3dwards** [reply]
I think you are looking at roughly $20 per month per load balancer + IP so not much.

I believe most providers smallest unit for billing would be an hour which would be pennies.

**Reply by josibake** [reply]
and for local dev, no extra cost, but just extra resources? I can't remember what it looks like to run a load balancer on Minikube, or rather I think Minikube has an ingress hack that doesn't need a load balancer?

**Reply by m3dwards** [reply]
Locally wouldn't be resource intensive. From the last time I did it (and things may have changed), on Docker Desktop it just magically worked and you can hit localhost. On Minikube you need to start a long running process called `minikube tunnel` to enable access to the load balancer.

**Reply by josibake** [reply]
haha minikube tunnel sounds sus

![](https://i.imgflip.com/934va5.jpg)

**Reply by josibake** [reply]
altho minikube tunnel would be a solid improvement over port forwarding if it allows us to get rid of stuff like this in our codebase: https://github.com/bitcoin-dev-project/warnet/pull/531#discussion_r1754595121

**Reply by m3dwards** [reply]
I can have a go at knocking up a proof of concept but I would prefer to do this after I know we are set for tabconf as I don't think this is a dealbreaker for that.

**Reply by josibake** [reply]
yeah, agree. mainly just want to keep track of this and keep it top of mind because im starting to see lots of tech debt related to port forwarding and id like to keep track of it so we can hopefully simplify all of this by just moving to ingress. worth mentioning, nothing stops users from port-forwarding on their machine when doing local dev! Id just strongly prefer if we weren't writing code to automatically start / stop port forwarding or having it bleed into our cli commands.

**Comment by josibake** [comment]
Hate to be on the warpath about this, but I seriously doubt port forwarding is going work for Tabconf. We are trying to build a "production" use case around a debugging feature, which is going to bite us in the ass. From reading the docs a bit, port-forwarding is meant for a single user of kubectl to quickly gain access to ports; I haven't seen anything that indicatese port-forwarding will work with `N` laptops all trying to forward to the same port on a cluster. I'm almost certain if we have 30+ laptops all trying to port forwarding to the same port on the cluster, it's going to fail.

if the idea is we will be the only ones running a dashboard, that might work but I think it will still be flakey and need us to be restarting the port forward. From what ive read, port forwarding is also slow and not meant for lots of data, which is terrible for a dashboard. Theres also nothing from stopping users from accidentally port-forwarding (e.g., via `warnet dashboard`) which could then take down our dashboard during the event.

If someone can prove me wrong on this, fantastic, I'll shut up about it! In the absence of that I think we need to do ingress. I can pick it up week after next, but in the meantime I'd suggest we stop building features around port forwarding. There is also nothing that stops us individual devs from port forwarding by just running the kubectl commands ourselves, so really no need for it to be in the python code.

**Reply by pinheadmz** [reply]
I think you are totally right about ingress being the correct thing here, but I also don't think everyone in the room at tabconf needs to watch grafana locally. There is supposedly one big screen for us to use as a scoreboard.

**Reply by josibake** [reply]
> but I also don't think everyone in the room at tabconf needs to watch grafana locally

I somewhat agree, but I don't see that as an argument to not do ingress? That just means we have ingress and the only person using it is the single browser we have open with the Caddy dashboard. You also didn't address any of the points I made regarding port-forwarding not being able to keep up with a sufficiently large dashboard, or supporting realtime updates. It also doesn't prevent anyone from nuking the dashboard by port-forwarding from their laptop. I almost guarantee someone will look at the commands in warnet , see `warnet dashboard` and be like huh, wonder what this does!

It's fine if you don't see these as actual concerns, and perhaps I'm being paranoid! But I think that still leaves the concern that even though almost everyone has said "Yeah, I agree ingress is the right way to go," we are still spending time writing and reviewing code that shoe-horns automatic port-forwarding into logging and caddy. So if we all agree ingress is the right way, let's just do ingress and stop adding tech debt to the codebase.

If instead people are happy with port-forwading, and that's why we keep writing code for it, then I'd really appreciate someone validating that the concerns I've raised are not actual concerns.

If we _still_ feel like ingress is the right way but we don't have time before Tabconf to do it, as a compromise I'd say lets rip out all the port-forwarding code and do it all manually. Since we will be the only ones running a dashboard, the users wont need to do anything manual , so we are all good.

**Reply by pinheadmz** [reply]
I agree with all your points, we should do it

**Comment by pinheadmz** [comment]
yeah port forwarding sucks. it dropped on my laptop and i dont know how to get it back using warnet without `deploy`

**Comment by m3dwards** [comment]
Merged in #607
